lib/scripts.js

```diff
@@ -301,200 +213,66 @@ var scripts = {
-    var lock = job.lock;
-    if (!lock) {
-      throw new Error('Unable to release nonexisting lock');
-    }
-    return lock.unlock()
+  takeLock: function(queue, job){
+    var lua = loadScriptIfNeeded(queue.client, 'takeLock');
+
+    return runScript(queue.client, 'takeLock', lua, [job.lockKey()], [queue.token, queue.LOCK_DURATION]);
   },
+
   /**
     It checks if the job in the top of the delay set should be moved back to the
     top of the  wait queue (so that it will be processed as soon as possible)
   */
   updateDelaySet: function(queue, delayedTimestamp){
-    var script = [
-      'local RESULT = redis.call("ZRANGE", KEYS[1], 0, 0, "WITHSCORES")',
-      'local jobId = RESULT[1]',
-      'local score = RESULT[2]',
-      'if (score ~= nil) then',
-      ' score = score / 0x1000 ',
-      ' if (score <= tonumber(ARGV[2])) then',
-      '  redis.call("ZREM", KEYS[1], jobId)',
-      '  redis.call("LREM", KEYS[2], 0, jobId)',
-      '  redis.call("LPUSH", KEYS[3], jobId)',
-      '  redis.call("PUBLISH", KEYS[4], jobId)',
-      '  redis.call("HSET", ARGV[1] .. jobId, "delay", 0)',
-      '  local nextTimestamp = redis.call("ZRANGE", KEYS[1], 0, 0, "WITHSCORES")[2]',
-      '  if(nextTimestamp ~= nil) then',
-      '   nextTimestamp = nextTimestamp / 0x1000',
-      '   redis.call("PUBLISH", KEYS[1], nextTimestamp)',
-      '  end',
-      '  return nextTimestamp',
-      ' end',
-      ' return score',
-      'end'].join('\n');
+    var lua = loadScriptIfNeeded(queue.client, 'updateDelaySet');
 
     var keys = _.map([
       'delayed',
       'active',
       'wait',
-      'jobs'], function(name){
+      'added'], function(name){
         return queue.toKey(name);
     });
 
-    var args = [
-      queue.client,
-      'updateDelaySet',
-      script,
-      keys.length,
-      keys[0],
-      keys[1],
-      keys[2],
-      keys[3],
-      queue.toKey(''),
-      delayedTimestamp
-    ];
+    var args = [queue.toKey(''), delayedTimestamp]; 
 
-    return execScript.apply(scripts, args);
+    return runScript(queue.client, 'updateDelaySet', lua, keys, args);
   },
 
   /**
-   * Looks for unlocked jobs in the active queue. There are two circumstances in which a job
-   * would be in 'active' but NOT have a job lock:
+   * Looks for unlocked jobs in the active queue. 
    *
-   *  Case A) The job was being worked on, but the worker process died and it failed to renew the lock.
+   *    The job was being worked on, but the worker process died and it failed to renew the lock.
    *    We call these jobs 'stalled'. This is the most common case. We resolve these by moving them
    *    back to wait to be re-processed. To prevent jobs from cycling endlessly between active and wait,
    *    (e.g. if the job handler keeps crashing), we limit the number stalled job recoveries to MAX_STALLED_JOB_COUNT.
-
-   *  Case B) The job was just moved to 'active' from 'wait' and the worker that moved it hasn't gotten
```

Comment:
You might want to remove `This is the most common case` above since it's now the only case.